galley/pkg/crd/validation/webhook.go

```diff
@@ -0,0 +1,354 @@
+	}
+	adapters := config.AdapterInfoMap(adapter.Inventory(), template.NewRepository(info).SupportsTemplate)
+
+	storeURL := fmt.Sprintf("k8s://%s?retry-timeout=%v", kubeconfig, 2*time.Second)
+
+	s, err := store.NewRegistry(config.StoreInventory()...).NewStore(storeURL)
+	if err != nil {
+		return nil, err
+	}
+	rv, err := validator.New(checker.NewTypeChecker(), "", s, adapters, templates)
+	if err != nil {
+		return nil, err
+	}
+	return store.NewValidator(rv, runtimeConfig.KindMap(adapters, templates)), nil
+}
+
+// NewWebhook creates a new instance of the admission webhook controller.
+func NewWebhook(p WebhookParameters) (*Webhook, error) {
+	pair, err := tls.LoadX509KeyPair(p.CertFile, p.KeyFile)
+	if err != nil {
+		return nil, err
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	// watch the parent directory of the target files so we can catch
+	// symlink updates of k8s secrets
+	for _, file := range []string{p.CertFile, p.KeyFile} {
+		watchDir, _ := filepath.Split(file)
+		if err := watcher.Watch(watchDir); err != nil {
+			return nil, fmt.Errorf("could not watch %v: %v", file, err)
+		}
+	}
+
+	wh := &Webhook{
+		server: &http.Server{
+			Addr: fmt.Sprintf(":%v", p.Port),
+		},
+		watcher:             watcher,
+		healthCheckInterval: p.HealthCheckInterval,
+		healthCheckFile:     p.HealthCheckFile,
+		certFile:            p.CertFile,
+		keyFile:             p.KeyFile,
+		cert:                &pair,
+		descriptor:          p.PilotDescriptor,
+		validator:           p.MixerValidator,
+	}
+
+	// mtls disabled because apiserver webhook cert usage is still TBD.
+	wh.server.TLSConfig = &tls.Config{GetCertificate: wh.getCert}
+	h := http.NewServeMux()
+	h.HandleFunc("/admitpilot", wh.serveAdmitPilot)
+	h.HandleFunc("/admitmixer", wh.serveAdmitMixer)
+	wh.server.Handler = h
+
+	return wh, nil
+}
+
+// Run implements the webhook server
+func (wh *Webhook) Run(stop <-chan struct{}) {
+	go func() {
+		if err := wh.server.ListenAndServeTLS("", ""); err != nil && err != http.ErrServerClosed {
+			log.Errorf("ListenAndServeTLS for admission webhook returned error: %v", err)
+		}
+	}()
+	defer wh.watcher.Close() // nolint: errcheck
+	defer wh.server.Close()  // nolint: errcheck
+
+	var healthC <-chan time.Time
+	if wh.healthCheckInterval != 0 && wh.healthCheckFile != "" {
+		t := time.NewTicker(wh.healthCheckInterval)
+		healthC = t.C
+		defer t.Stop()
+	}
+	var timerC <-chan time.Time
+
+	for {
+		select {
+		case <-timerC:
+			pair, err := tls.LoadX509KeyPair(wh.certFile, wh.keyFile)
+			if err != nil {
+				log.Errorf("reload cert error: %v", err)
+				break
+			}
+			wh.mu.Lock()
+			wh.cert = &pair
+			wh.mu.Unlock()
+		case event := <-wh.watcher.Event:
+			// use a timer to debounce cert updates
+			if event.IsModify() || event.IsCreate() {
+				timerC = time.After(watchDebounceDelay)
+			}
+		case err := <-wh.watcher.Error:
+			log.Errorf("Watcher error: %v", err)
+		case <-healthC:
```

Comment:
Why do we need the health check file here? Is this needed for web hooks in general?

Reply:
All Istio components should have health checks. https://goo.gl/QRRdwm recommends that components use file based health checking.